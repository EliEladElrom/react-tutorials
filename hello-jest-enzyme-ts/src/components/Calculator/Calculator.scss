.calculator-output {
  background: url('/calculator-input.jpg') no-repeat right top;
  color: #fff;
  font-size: 50px;
  width: 464px;
  height: 120px;
  text-align: center;
  display: flex;
  justify-content: center;
  align-items: center;
}
