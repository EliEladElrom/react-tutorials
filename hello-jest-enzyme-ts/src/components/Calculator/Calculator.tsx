/*
Author: Eli Elad Elrom
Website: https://EliElrom.com
License: MIT License
Component: src/components/Calculator/Calculator.tsx
*/

import React from 'react'
import './Calculator.scss'
// @ts-ignore
import ImageMapper from 'react-image-mapper'

export const URL = 'calculator.jpg'
export const MAP = {
  name: 'my-map',
  areas: [
    {
      name: '0',
      shape: 'rect',
      coords: [3, 387, 227, 474],
    },
    {
      name: '1',
      shape: 'rect',
      coords: [2, 291, 112, 382],
    },
    {
      name: '2',
      shape: 'rect',
      coords: [116, 290, 227, 382],
    },
    {
      name: '3',
      shape: 'rect',
      coords: [342, 382, 232, 290],
    },
    {
      name: '4',
      shape: 'rect',
      coords: [3, 194, 111, 290],
    },
    {
      name: '5',
      shape: 'rect',
      coords: [115, 193, 227, 290],
    },
    {
      name: '6',
      shape: 'rect',
      coords: [231, 194, 343, 290],
    },
    {
      name: '7',
      shape: 'rect',
      coords: [4, 97, 111, 191],
    },
    {
      name: '8',
      shape: 'rect',
      coords: [115, 99, 227, 191],
    },
    {
      name: '9',
      shape: 'rect',
      coords: [231, 98, 343, 191],
    },
    {
      name: '+',
      shape: 'rect',
      coords: [348, 291, 463, 382],
    },
    {
      name: '-',
      shape: 'rect',
      coords: [348, 195, 463, 290],
    },
    {
      name: '*',
      shape: 'rect',
      coords: [348, 98, 463, 191],
    },
    {
      name: '/',
      shape: 'rect',
      coords: [348, 3, 463, 93],
    },
    {
      name: '=',
      shape: 'rect',
      coords: [348, 387, 463, 474],
    },
  ],
}

export default class Calculator
  extends React.Component<ICalculatorProps, ICalculatorState>
  implements ICalculator {
  constructor(props: ICalculatorProps) {
    super(props)
    this.state = {
      output: 0,
      operatorType: '',
      number1: 0,
      number2: -1,
    }
  }
  startOver = () => {
    this.setState({
      operatorType: '',
      number1: 0,
      number2: -1,
      output: 0,
    })
  }
  calculateTwoNumbers = (num1: number, num2: number, operator: string) => {
    let retVal = 0
    switch (operator) {
      case '+':
        retVal = num1 + num2
        break
      case '-':
        retVal = num1 - num2
        break
      case '*':
        retVal = num1 * num2
        break
      case '/':
        retVal = num1 / num2
        break
      default:
        alert('Operator not recognized')
    }
    return retVal
  }
  clicked = (btnName: string) => {
    switch (btnName) {
      case '-':
      case '*':
      case '/':
      case '+':
        this.setState({
          operatorType: btnName,
          number1: Number(this.state.output),
        })
        break
      case '=':
        this.setState({
          output: this.calculateTwoNumbers(
            this.state.number1,
            Number(this.state.output),
            this.state.operatorType
          ),
        })
        break
      default:
        let isFirstDigitNumber2 = this.state.operatorType && this.state.number2 === -1
        this.setState({
          number2: isFirstDigitNumber2 ? 0 : this.state.number2,
          output: isFirstDigitNumber2 ? Number(btnName) : Number(this.state.output + btnName),
        })
    }
  }
  render() {
    return (
      <>
          <a href="http://twitter.com/elieladelrom" className="follow">
            @elieladelrom
          </a>
          <h1 className="title">
            {this.props.componentTitle} - Version #{this.props.version}
          </h1>
          <p>
            <button id="btn" onClick={this.startOver}>Start Over</button>
          </p>
          <div className="calculator-output">{this.state.output}</div>
          <ImageMapper src={URL} map={MAP} onClick={(area: any) => { this.clicked(area.name) } } />
      </>
    )
  }
}

export interface ICalculator extends React.Component<ICalculatorProps, ICalculatorState> {
  startOver?(): void
  calculateTwoNumbers(num1: number, num2: number, operator: string): void
  clicked(area: any): void
}

interface ICalculatorProps {
  componentTitle: string
  version: string
}

interface ICalculatorState {
  output: number
  operatorType: string
  number1: number
  number2: number
}
