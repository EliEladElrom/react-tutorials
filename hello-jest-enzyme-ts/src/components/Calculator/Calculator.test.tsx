// src/components/Calculator/Calculator.test.tsx

import React from 'react'
import { shallow, mount } from 'enzyme'
import Calculator, {ICalculator} from './Calculator'

// non-interactive components - using it (single test)
it('should render the link url', () => {
  const wrapper = shallow(
      <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
  )
  const a = wrapper.find('a')
  const result = a.text()

  expect(result).toBe('@elieladelrom')
})

// describe(name, fn) creates a block that groups together several related tests in one "test suite".
describe('`Calculator Snapshots', () => {
  it('should render our Snapshots correctly', () => {
    const wrapper = shallow(
        <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})

// it(is aliased by test so it does the same thing as it)
test('should render component title', () => {
  const wrapper = shallow(
      <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
  )
  const title = wrapper.find('h1.title').text()

  expect(title).toBe('Online `Calculator - Version #0.01-beta')
})

test('Testing output indirectly - should clean our result box clicking clear', () => {
  const wrapper = shallow(
      <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
  )
  const btn = wrapper.find('#btn')
  btn.simulate('click')

  const output_2 = wrapper.find('.calculator-output').text()
  expect(output_2).toBe('0')
})

describe('Testing `Calculator calculateTwoNumbers testsuite directly', () => {
  test('Testing calculateTwoNumbers Directly - add', () => {
    const wrapper = shallow(
        <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
    )
    const instance = wrapper.instance() as ICalculator
    expect(instance.calculateTwoNumbers(1, 2, '+')).toBe(3)
  })
  test('Testing calculateTwoNumbers Directly - multiple', () => {
    const wrapper = shallow(
        <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
    )
    const instance = wrapper.instance() as ICalculator
    expect(instance.calculateTwoNumbers(2, 2, '*')).toBe(4)
  })
})

test('test clicked calculator button method', () => {
  const wrapper = shallow(
      <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
  )
  const instance = wrapper.instance() as ICalculator
  instance.clicked('1')
  expect(wrapper.state('output')).toBe(1)
})

test('spy', () => {
  const wrapper = shallow(
      <Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />
  )
  const instance = wrapper.instance() as ICalculator
  jest.spyOn(instance, 'startOver')
  wrapper.find('button').simulate('click')
  expect(wrapper.state('output')).toBe(0)
})

import sinon from 'sinon'

describe('Loader component', () => {
  it('should render complete after x seconds', () => {
    const wrapper = shallow(<Calculator componentTitle={'Online `Calculator'} version={'0.01-beta'} />)
    let clock = sinon.useFakeTimers()
    const instance = wrapper.instance() as ICalculator
    instance.startLoader(3000)
    clock.tick(3000)

    const title = wrapper.find('h1.subTitle').text()
    expect(title).toBe('Loading Complete')
  })
})
