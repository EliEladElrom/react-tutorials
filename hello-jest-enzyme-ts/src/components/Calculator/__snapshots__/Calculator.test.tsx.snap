// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Calculator Snapshots should render our Snapshots correctly 1`] = `
<Fragment>
  <a
    className="follow"
    href="http://twitter.com/elieladelrom"
  >
    @elieladelrom
  </a>
  <h1
    className="title"
  >
    Online \`Calculator
     - Version #
    0.01-beta
  </h1>
  <p>
    <button
      id="btn"
      onClick={[Function]}
    >
      Start Over
    </button>
  </p>
  <div
    className="calculator-output"
  >
    0
  </div>
  <ImageMapper
    active={true}
    fillColor="rgba(255, 255, 255, 0.5)"
    lineWidth={1}
    map={
      Object {
        "areas": Array [
          Object {
            "coords": Array [
              3,
              387,
              227,
              474,
            ],
            "name": "0",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              2,
              291,
              112,
              382,
            ],
            "name": "1",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              116,
              290,
              227,
              382,
            ],
            "name": "2",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              342,
              382,
              232,
              290,
            ],
            "name": "3",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              3,
              194,
              111,
              290,
            ],
            "name": "4",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              115,
              193,
              227,
              290,
            ],
            "name": "5",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              231,
              194,
              343,
              290,
            ],
            "name": "6",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              4,
              97,
              111,
              191,
            ],
            "name": "7",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              115,
              99,
              227,
              191,
            ],
            "name": "8",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              231,
              98,
              343,
              191,
            ],
            "name": "9",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              291,
              463,
              382,
            ],
            "name": "+",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              195,
              463,
              290,
            ],
            "name": "-",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              98,
              463,
              191,
            ],
            "name": "*",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              3,
              463,
              93,
            ],
            "name": "/",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              387,
              463,
              474,
            ],
            "name": "=",
            "shape": "rect",
          },
        ],
        "name": "my-map",
      }
    }
    onClick={[Function]}
    src="calculator.jpg"
    strokeColor="rgba(0, 0, 0, 0.5)"
  />
</Fragment>
`;

exports[`CalculatorJS Snapshots should render our Snapshots correctly 1`] = `
<Fragment>
  <a
    className="follow"
    href="http://twitter.com/elieladelrom"
  >
    @elieladelrom
  </a>
  <h1
    className="title"
  >
    Online CalculatorJS
     - Version #
    0.01-beta
  </h1>
  <p>
    <button
      onClick={[Function]}
    >
      Start Over
    </button>
  </p>
  <div
    className="calculator-output"
  >
    0
  </div>
  <ImageMapper
    active={true}
    fillColor="rgba(255, 255, 255, 0.5)"
    lineWidth={1}
    map={
      Object {
        "areas": Array [
          Object {
            "coords": Array [
              3,
              387,
              227,
              474,
            ],
            "name": "0",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              2,
              291,
              112,
              382,
            ],
            "name": "1",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              116,
              290,
              227,
              382,
            ],
            "name": "2",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              342,
              382,
              232,
              290,
            ],
            "name": "3",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              3,
              194,
              111,
              290,
            ],
            "name": "4",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              115,
              193,
              227,
              290,
            ],
            "name": "5",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              231,
              194,
              343,
              290,
            ],
            "name": "6",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              4,
              97,
              111,
              191,
            ],
            "name": "7",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              115,
              99,
              227,
              191,
            ],
            "name": "8",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              231,
              98,
              343,
              191,
            ],
            "name": "9",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              291,
              463,
              382,
            ],
            "name": "+",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              195,
              463,
              290,
            ],
            "name": "-",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              98,
              463,
              191,
            ],
            "name": "*",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              3,
              463,
              93,
            ],
            "name": "/",
            "shape": "rect",
          },
          Object {
            "coords": Array [
              348,
              387,
              463,
              474,
            ],
            "name": "=",
            "shape": "rect",
          },
        ],
        "name": "my-map",
      }
    }
    onClick={[Function]}
    src="calculator.jpg"
    strokeColor="rgba(0, 0, 0, 0.5)"
  />
</Fragment>
`;
